### ISSUE TITLE
Feature Enhancement - Calling back into an anonymous JS function

### ISSUE THREAD

**Issue by rassilon** [cefsharp/CefSharp#825]
Currently, `IRenderProcess` only supports calling globally exposed JS functions. That doesn't mesh very well with modern JS programming styles that use oodles of anonymous functions.

Here is a very vague outline of how this might happen in CefSharp:

`TypeUtils.cpp:ConvertFromCEF`: If V8 function: then track function via map approach suggested by [CEF](https://code.google.com/p/chromiumembedded/wiki/GeneralUsage#Custom_Implementation), serialize via browser_id, frameId, callbackId back to browser process wrapper class. `JavascriptCallbackWrapper` or something. I'm not entirely sure which class the suggested map should exist in CefSharp though. TypeUtils is a static class currently. `CBrowserWrapper` maybe?

`IRenderProcess`: Add `IAsyncResult BeginExecuteCallback(callbackStateObject, object[] args)` and `EndExecuteCallback()` methods similar to existing script executions methods on `IRenderProcess`.

`CefAppUnmanagedWrapper::OnContextReleased`: Release the new map.

NOTE: this wouldn't try to do anything to track the JavaScript `this` value, that'd be up to the callback to supply the appropriate bind call.

Thoughts?
Bill

**Comment by amaitland** [comment]
> TypeUtils is a static class currently. CBrowserWrapper maybe?

`TypeUtils` is just a helper class, being used in a few place, I'd try and void changing it if possible. I think `CefBrowserWrapper` is the place to keep the dictionary/map.

On the `C++` side of things there's probably some refactoring that needs to happen to implement this cleanly, that will be the messiest part.

For me the key parts are:
- Iterate through the params, convert where possible, if a function then generate a unique Id, store a reference to the context and function it's self in the dictionary/map. Insert a proxy object into the params passed to the `WCF` call. 
- When `OnContextReleased` then release the map contained within the browser wrapper.
- Make appropriate `WCF` calls

Perhaps roughly map out the public facing `API`? 

```c#
public string Repeat(string str, int n, IJavascriptCallback callBack)
{

}
```

The slightly less intuitive approach would be to have a method on the browser to execute the callback, I like this idea from the point of view the callback wrapper is just a lightweight object containing an `id` plus some other information, no need to keep track of a `WCF` proxy, so I'd say there will be fewer memory leaks. The browser actually executes the calls.
`browser.ExecuteJavascriptCallBack(callBack, params);`

The other more traditional approach is of course something like
`callBack.Execute(params);`

From an implementation point of view this is a little more complicated as references to the proxy need to be maintained and responsibly disposed of.

Do we need a cancel method to cleanup? For long running single page applications this might be necessary.

Are you planning to implement this in the near future?

**Comment by rassilon** [comment]
I'm not planning on working on it. Just thought I should jot the thought down somewhere. 


Bill

**Comment by amaitland** [comment]
No problem :+1:

**Comment by arsher** [comment]
Hi,

I actually started working on this a few weeks ago, but didn't have the time to finish it. I tried a few approaches but I always ran into the issue of getting rid of the unneeded references of the callbacks. I couldn't find a way to determine if an instance of CefV8Value belongs to a previously mapped function so I always create a new ref (new unique id etc.) which is not the best idea, for example if the same callback is passed multiple times in a long loop...Do you guys have any idea to circumvent this?

Also on the API side what I had in mind is to create strong typed delegates for the passed functions (in the best case scenario that's what I want to provide to my users), something like this:

```
class BoundObject
{
    public void RegisterCallback(Func<int, int, int> callback)
    {
        ...
    }
}
```

But this approach raises the question of cleanup again...I was thinking adding a DynamicMethod on an instance of an object (one per delegate type or one per callback) and adding some cleanup code in the finalizer of that object. But I'm still experimenting, it might be just easier and cleaner to have something like the above suggested  `IJavascriptCallback`.

Cheers,
David

**Comment by amaitland** [comment]
@arsher Awesome :+1: I haven't really looked at any of the ideas in details, so perhaps it would be productive to push the progress you have made and we can collectively discuss inline?

> But I'm still experimenting, it might be just easier and cleaner to have something like the above suggested IJavascriptCallback.

Having a strongly typed `Func` sounds like a fantastic feature to have, my biggest concern is of course memory leaks. It may simply be as simple as keeping a `WeakReference` to the `WCF` proxy behind the scenes, have to look at it further.

In the short term, I'd personally implement the simplest approach and make iterative improvements from there. 

Anyways, great news on your progress. Let's keep the dialog open :+1:

**Comment by rassilon** [comment]
I would leave strongly typed wrapper until after the basic idea is working. 

> re: IJavascriptCallback and cleanup

I think a `WeakReference` to the `WCF` proxy makes sense. We'd also need a `RelaseCallbackContext` method or something on `IRenderProcess` to perform a manual context release if the data is still in the render side context mapping.

`IJavascriptCallback` should probably inherit from `IDisposable` and have a C# `~JavascriptCallback` finalizer as well. Just need to be careful about exceptions occurring during finalization / race conditions involved in the `WCF` shutdown process. 

Bill

**Comment by arsher** [comment]
Hey guys,

I finally had some time to turn my stuff into a few commits and also to add the ```IJavascriptCallback ``` approach. So I have a prototype on a branch here: https://github.com/arsher/CefSharp/tree/jscallback

David

**Comment by amaitland** [comment]
@arsher Fantastic :smile:  If you submit a `PR` we can continue the conversation in context :+1:

**Comment by amaitland** [comment]
Merged #887 (which included the fine work from @arsher in PR #868)
